README.md
templer.affinitic
=================

http://www.affinitic.be/
info@affinitic.be

Summary
-------

The package templer.affinitic allow to create new packages the simpliest possible and corresponding to Affinitic needs.


Available templates
-------------------

Affinitic templer templates :

* affinitic\_basic\_plone\_buildout : A basic Plone buildout skeleton
* affinitic\_diazo\_theme : A basic Diazo theme package skeleton (with buildout)


Installation
------------

To use this product, you need to clone and install his buildout:

        cd buildouts
        git clone git@github.com:affinitic/templer.affinitic.git
        cd templer.affinitic
        virtualenv -p python2.6 --no-site-packages .
        bin/python bootstrap.py
        bin/buildout

Now that the templer.affinitic tool is installed, the script allowing you to create new package is:

        bin/templer

Do not hesitate to make an alias of the script to use it easily use it anywhere on your computer.


Script usage
------------

The script needs a template as argument, the list of available templates is showed when launching the script without argument or with '--list'

        bin/templer
        bin/templer --list

This will show:

        Affinitic

        affinitic_basic_plone_buildout: A basic Plone buildout skeleton
        affinitic_diazo_theme:          A basic Diazo theme package skeleton

... and some other zope/plone templates. We are of course interested by the Affinitic's ones.

To finally create a new package, you first need to be in the folder where you want the package to be created:

        cd buildouts

And to execute the script with the desired template:

        ../templer.affinitic/bin/templer affinitic_diazo_theme

You can also pass the name of the project in the arguments:

        ../templer.affinitic/bin/templer affinitic_diazo_theme my.package

Next, just answer the questions. They are the most simplified with good default answers:

        Enter project name: my.package
        Namespace Package Name (Name of outer namespace package) ['my']:
        Package Name (Name of the inner namespace package) ['package']:
        Web hosting service (If public, choose github. If private, choose bitbucket (github/bitbucket)) ['github']:
        Description (One-line description of the project) ['']: Description de mon package
        Project URL (URL of the homepage for this project) ['https://github.com/affinitic/my.package']:
        Zope port (Specify a port for your zope) ['8080']:
        Creating directory ./my.package

For example, this template will create a plone with a diazo and it buildout.

There is still some issues, you can follow them in the github issues of the package.


docs/HISTORY.txt
Changelog
=========

0.1 (unreleased)
----------------

- Nothing changed yet


src/templer/affinitic/tests/test_templates.py
# -*- coding: utf-8 -*-

import unittest2 as unittest

import os

from templer.core.tests.test_templates import doc_suite

current_dir = os.path.abspath(os.path.dirname(__file__))


def test_suite():
    """returns the test suite"""
    suite = doc_suite(current_dir)
    return suite

if __name__ == '__main__':
    unittest.main(defaultTest='test_suite')


src/templer/affinitic/base.py
# -*- coding: utf-8 -*-
"""
templer.affinitic

Licensed under the GPL license, see LICENCE.txt for more details.
Copyright by Affinitic sprl
"""


from templer.core.basic_namespace import BasicNamespace
from templer.core.base import get_zopeskel_prefs, wrap_help_paras
from templer.core.create import NoDefault, BadCommand
from templer.core.vars import (
    ValidationException,
    StringVar,
    StringChoiceVar)
from templer.core.base import get_var

from textwrap import TextWrapper


class AffiniticBaseTemplate(BasicNamespace):

    use_cheetah = True

    def __init__(self, *args, **kw):
        super(AffiniticBaseTemplate, self).__init__(*args, **kw)

        # Add new var
        self.vars.insert(3, StringChoiceVar(
            'hoster',
            title='Web hosting service',
            description='If public, choose github. If private, choose bitbucket (github/bitbucket)',
            default='github',
            choices=('github', 'bitbucket'),
        ))

        # Change existing vars
        get_var(self.vars, 'version').default = '0.1'
        get_var(self.vars, 'version').questionable = False

        get_var(self.vars, 'author').default = 'Affinitic'
        get_var(self.vars, 'author').questionable = False

        get_var(self.vars, 'author_email').default = 'info@affinitic.be'
        get_var(self.vars, 'author_email').questionable = False

        get_var(self.vars, 'description').required = True

        get_var(self.vars, 'long_description').questionable = False

        get_var(self.vars, 'keywords').default = 'Affinitic'
        get_var(self.vars, 'keywords').questionable = False

        get_var(self.vars, 'license_name').questionable = False

        get_var(self.vars, 'zip_safe').questionable = False

        get_var(self.vars, 'url').default = 'https://github.com/affinitic/'

        # Remove unwanted vars
        for var in self.vars[:]:
            if var.name == 'expert_mode':
                self.vars.remove(var)

    def check_vars(self, vars, cmd):
        """
        Override of templer.core.base.BaseTemplate.check_vars

        Add the possibility to specify if a variable will
          be questioned in the commande prompt or not
        eg:
            class MyTemplate(AffiniticTemplate):
                vars = copy.deepcopy(BasicNamespace.vars)
                get_var(vars, 'version').questionable = False
        """
        if not hasattr(cmd, '_deleted_once'):
            del vars['package']
            cmd._deleted_once = True

        # if we need to notify users of anything before they start this
        # whole process, we can do it here.
        self.print_zopeskel_message('pre_run_msg')

        # Copied and modified from PasteScript's check_vars--
        # the method there wasn't hookable for the things
        # we need -- question posing, validation, etc.
        #
        # Admittedly, this could be merged into PasteScript,
        # but it was decided it was easier to limit scope of
        # these changes to ZopeSkel, as other projects may
        # use PasteScript in very different ways.

        cmd._deleted_once = 1      # don't re-del package

        textwrapper = TextWrapper(
            initial_indent="|  ",
            subsequent_indent="|  ",
        )

        # now, mostly copied direct from paster
        expect_vars = self.read_vars(cmd)
        if not expect_vars:
            # Assume that variables aren't defined
            return vars
        converted_vars = {}
        errors = []

        config = get_zopeskel_prefs()
        # pastescript allows one to request more than one template (multiple
        # -t options at the command line) so we will get a list of templates
        # from the cmd's options property
        requested_templates = cmd.options.templates
        for var in expect_vars:
            for template in requested_templates:
                if config.has_option(template, var.name):
                    var.default = config.get(template, var.name)
                    break
            else:
                # Not found in template section, now look explicitly
                # in DEFAULT section
                if config.has_option('DEFAULT', var.name):
                    var.default = config.get('DEFAULT', var.name)

        self.override_package_names_defaults(vars, expect_vars)
        unused_vars = vars.copy()

        for var in expect_vars:
            response = self.null_value_marker
            if var.name not in unused_vars:
                if cmd.interactive \
                   and (not hasattr(var, 'questionable')
                        or var.questionable == True):
                    prompt = var.pretty_description()
                    while response is self.null_value_marker:
                        response = cmd.challenge(prompt, var.default,
                                                 var.should_echo)
                        if response == '?':
                            help = var.further_help().strip() % converted_vars
                            print
                            wrap_help_paras(textwrapper, help)
                            print
                            response = self.null_value_marker

                        if hasattr(var, 'required') \
                           and var.required == True \
                           and response.strip() == '':
                            print
                            print "You must answer to this question!"
                            print
                            response = self.null_value_marker

                        if response is not self.null_value_marker:
                            try:
                                response = var.validate(response)
                            except ValidationException, e:
                                print e
                                response = self.null_value_marker
                elif var.default is NoDefault:
                    errors.append('Required variable missing: %s'
                                  % var.full_description())
                else:
                    response = var.validate(var.default)
            else:
                response = var.validate(unused_vars.pop(var.name))

            converted_vars[var.name] = response
            # if a variable has structures associated, we will insert them
            # in the template required_structures property at this time, let's
            # test first to see if we need to do anything.
            if var._is_structural:
                self._set_structure_from_var(var, str(response))

            # filter the vars for mode.
            if var.name == 'expert_mode':
                expert_mode = converted_vars['expert_mode']
                hidden = self._filter_for_modes(expert_mode, expect_vars)
                unused_vars.update(hidden)

            self.process_dependents_vars(vars, var, response, expect_vars)

        if errors:
            raise BadCommand(
                'Errors in variables:\n%s' % '\n'.join(errors))
        converted_vars.update(unused_vars)
        vars.update(converted_vars)

        result = converted_vars

        return result

    def process_dependents_vars(self, vars, changed_var, changed_var_value, expect_vars):
        """
        Process some changes on vars that need to know the value of precedent var
        """
        if changed_var.name == 'hoster':
            if changed_var_value == 'github':
                host = 'github.com'
            elif changed_var_value == 'bitbucket':
                host = 'bitbucket.org'
            url = 'https://%s/affinitic/%s' % (host, vars['project'])
            get_var(expect_vars, 'url').default = url


class AffiniticBaseBuildoutTemplate(AffiniticBaseTemplate):
    category = "Affinitic"
    required_templates = []
    default_required_structures = ['bootstrap', 'gitignore']

    def __init__(self, *args, **kw):
        super(AffiniticBaseBuildoutTemplate, self).__init__(*args, **kw)

        # Add new var
        self.vars.append(StringVar(
            'port',
            title='Zope port',
            description='Specify a port for your zope',
            default='8080',
        ))
